src/dyncfg/src/lib.rs

```diff
@@ -242,45 +229,123 @@ impl ConfigEntry {
-    /// A 'Duration' shared value.
     Duration(Arc<RwLock<Duration>>),
 }
 
+impl From<ConfigVal> for ConfigValAtomic {
+    fn from(val: ConfigVal) -> ConfigValAtomic {
+        match val {
+            ConfigVal::Bool(x) => ConfigValAtomic::Bool(Arc::new(AtomicBool::new(x))),
+            ConfigVal::U32(x) => ConfigValAtomic::U32(Arc::new(AtomicU32::new(x))),
+            ConfigVal::Usize(x) => ConfigValAtomic::Usize(Arc::new(AtomicUsize::new(x))),
+            ConfigVal::OptUsize(x) => ConfigValAtomic::OptUsize(Arc::new(RwLock::new(x))),
+            ConfigVal::String(x) => ConfigValAtomic::String(Arc::new(RwLock::new(x))),
+            ConfigVal::Duration(x) => ConfigValAtomic::Duration(Arc::new(RwLock::new(x))),
+        }
+    }
+}
+
+impl ConfigValAtomic {
+    fn load(&self) -> ConfigVal {
+        match self {
+            ConfigValAtomic::Bool(x) => ConfigVal::Bool(x.load(SeqCst)),
+            ConfigValAtomic::U32(x) => ConfigVal::U32(x.load(SeqCst)),
+            ConfigValAtomic::Usize(x) => ConfigVal::Usize(x.load(SeqCst)),
+            ConfigValAtomic::OptUsize(x) => ConfigVal::OptUsize(*x.read().expect("lock poisoned")),
+            ConfigValAtomic::String(x) => {
+                ConfigVal::String(x.read().expect("lock poisoned").clone())
+            }
+            ConfigValAtomic::Duration(x) => ConfigVal::Duration(*x.read().expect("lock poisoned")),
+        }
+    }
+
+    fn store(&self, val: ConfigVal) {
+        match (self, val) {
+            (ConfigValAtomic::Bool(x), ConfigVal::Bool(val)) => x.store(val, SeqCst),
+            (ConfigValAtomic::U32(x), ConfigVal::U32(val)) => x.store(val, SeqCst),
+            (ConfigValAtomic::Usize(x), ConfigVal::Usize(val)) => x.store(val, SeqCst),
+            (ConfigValAtomic::OptUsize(x), ConfigVal::OptUsize(val)) => {
+                *x.write().expect("lock poisoned") = val
+            }
+            (ConfigValAtomic::String(x), ConfigVal::String(val)) => {
+                *x.write().expect("lock poisoned") = val
+            }
+            (ConfigValAtomic::Duration(x), ConfigVal::Duration(val)) => {
+                *x.write().expect("lock poisoned") = val
+            }
+            (ConfigValAtomic::Bool(_), val)
+            | (ConfigValAtomic::U32(_), val)
+            | (ConfigValAtomic::Usize(_), val)
+            | (ConfigValAtomic::OptUsize(_), val)
+            | (ConfigValAtomic::String(_), val)
+            | (ConfigValAtomic::Duration(_), val) => {
+                panic!("attempted to store {val:?} value in {self:?} parameter")
+            }
+        }
+    }
+}
+
 impl ConfigUpdates {
-    /// Adds the current value of the given config to this set of updates.
+    /// Adds an update for the given config and value.
     ///
     /// If a value of the same config has previously been added to these
     /// updates, replaces it.
-    pub fn add(&mut self, config: &ConfigEntry) {
+    pub fn add<T>(&mut self, config: &Config<T>, val: T)
```

Comment:
nit: when inlining the where in the type params doesn't cause it to wrap, I tend to do it (for information density reasons). i.e. `pub fn add<T: ConfigType>(&mut self, config: &Config<T>, val: T) {`

Reply:
Ah, sorry I lost track of this comment from the original PR. I thought I had addressed this but I must have lost it in a rebase.

FWIW I've actually followed the opposite style in all the Rust code I've written for Materialize, which is to always use `where` clauses, even for short bounds. It avoids ever having to make a judgment call as to what's short enough to inline, and protects against lazy refactors where new bounds repeatedly get stuffed into the inline position until it becomes unreadable. I've switched this over to the persist/dyncfg style, but we (i.e., Materialize eng broadly) may want to consider agreeing on an Official Position and adding it to STYLE.md. Really wish `rustfmt` would handle switching between the forms automatically; then I'd have no qualms about "inline if short" as the style, because we'd be deferring the judgment call to rustfmt.